### PR TITLE
KOGITO-556 - rest server for handling process instance error

### PR DESCRIPTION
**Pull request by Sara4994**

### CODE REVIEW

File: packages/management-console/server/restServer.js [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        const error = processId[0].instances.filter((err) => {
+            return err.processInstanceId === req.params.processInstanceId
+        })
+        res.send(error[0].error)
+    }
+}
+
+app.use(cors())
+// http://localhost:8090/management/process/{processId}/instances/{processInstanceId}/error
+app.get('/management/process/:processId/instances/:processInstanceId/error', controller.showError)
```
Comment (cristianonicolai) [inline packages/management-console/server/restServer.js:117]: how about retrigger and skip?
{processId}/instances/{processInstanceId}/retrigger
{processId}/instances/{processInstanceId}/skip

File: packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx [hunk 2]
```diff
@@ -50,6 +56,13 @@ const DataListItemComponent: React.FC<IOwnProps> = ({ id, instanceID, instanceSt
       }
     }
   `;
+  const handleModal = useCallback(async (_processID, _instanceID) => {
+    const processInstanceId = instanceID;
+    const processId = processID;
+    setOpenModal(true);
+    const result = await axios.get(`http://localhost:8090/management/process/${processId}/instances/${processInstanceId}/error`);
```
Comment (cristianonicolai) [inline packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx]: http://localhost:8090 should be replaced by the _endpoint_ attribute from the process instance.

File: packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx [hunk 3]
```diff
@@ -121,16 +139,49 @@ const DataListItemComponent: React.FC<IOwnProps> = ({ id, instanceID, instanceSt
...
               onSelect={onSelect}
               toggle={<KebabToggle onToggle={onToggle} />}
               dropdownItems={[
-                <DropdownItem key={1}>Abort</DropdownItem>,
+                <DropdownItem key={1}>Retry</DropdownItem>,
+                <DropdownItem key={2}>Skip</DropdownItem>,
+                <DropdownItem key={3} onClick = {() => handleModal(processID, instanceID) }>View Error</DropdownItem>,
               ]}
-            />
+            />: <Dropdown
```
Comment (cristianonicolai) [inline packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx]: actually, we dont have any actions now if not in error state. If in here should remove/hide kebab.

**Review by cristianonicolai — CHANGES_REQUESTED** [review]
@Sara4994 I believe it is missing the calls for skip and retry REST endpoints.

File: packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx [hunk 4]
```diff
@@ -121,16 +202,49 @@ const DataListItemComponent: React.FC<IOwnProps> = ({ id, instanceID, instanceSt
...
+            isPlain
+            position={DropdownPosition.right}
+            isOpen={isOpen}
+            onSelect={onSelect}
+            toggle={<KebabToggle onToggle={onToggle} />}
+            dropdownItems={[
+              <DropdownItem key={1}>Abort</DropdownItem>,
```
Comment (cristianonicolai) [inline packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx]: Abort option can be removed entirely for now.

File: packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx (hunk 3 above)
Comment (cristianonicolai) [reply packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx]: the entire kebab could be hidden in this case.

File: packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx [hunk 5]
```diff
@@ -27,15 +31,25 @@ export interface IOwnProps {
...
   const [isOpen, setisOpen] = useState(false);
   const [isLoaded, setisLoaded] = useState(false);
   const [isChecked, setisChecked] = useState(false);
   const [childList, setchildList] = useState([]);
+  const [openModal, setOpenModal] = useState(false);
+  const [skipAlertVisible, setSkipAlertVisible] = useState(false);
+  const [retryAlertVisible, setRetryAlertVisible] = useState(false);
+  const [error, setError] = useState('')
+  const [skipStatus, setSkipStatus] = useState('');
+  const [retryStatus, setretryStatus] = useState('');
```
Comment (cristianonicolai) [inline packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx]:
```suggestion
  const [retryStatus, setRetryStatus] = useState('');
```

File: packages/management-console/server/restServer.js [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+    },
+    callSkip: (req,res) => {
+        const {process} = mockData.management;
+        const processId = process.filter((data)     => {
+            return data.processId === req.params.processId
+        })
+        const error = processId[0].instances.filter((err) => {
+            return err.processInstanceId === req.params.processInstanceId
+        })
+        res.send(error[0].skip)
```
Comment (cristianonicolai) [inline packages/management-console/server/restServer.js]: we actually need that some of the request to fail, by that it needs to send a 401, 404, 500, 403 errors

File: packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx [hunk 7]
```diff
@@ -77,8 +125,41 @@ const DataListItemComponent: React.FC<IOwnProps> = ({ id, instanceID, instanceSt
       setisLoaded(true);
     }
   };
+
   return (
     <React.Fragment>
+      {skipAlertVisible && skipStatus === 'success' && (
+        <Alert
+            variant="success"
+            title="Success"
```
Comment (cristianonicolai) [inline packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx]:
```suggestion
            title="Skip operation"
```

File: packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx [hunk 8]
```diff
@@ -77,8 +125,41 @@ const DataListItemComponent: React.FC<IOwnProps> = ({ id, instanceID, instanceSt
...
+            action={<AlertActionCloseButton onClose={() => closeSkipAlert()} />}
+          >
+            Skip status : {skipStatus}
```
Comment (cristianonicolai) [inline packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx]: Process execution has successfully skipped node which was in error state.

File: packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx [hunk 9]
```diff
@@ -77,8 +125,41 @@ const DataListItemComponent: React.FC<IOwnProps> = ({ id, instanceID, instanceSt
...
+          </Alert>)}
+      {skipAlertVisible && skipStatus === 'failed' && (    
+          <Alert
+          variant="danger"
+          title="Failed"
```
Comment (cristianonicolai) [inline packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx]:
```suggestion
          title="Skip operation"
```

File: packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx [hunk 10]
```diff
@@ -77,8 +125,41 @@ const DataListItemComponent: React.FC<IOwnProps> = ({ id, instanceID, instanceSt
...
+          action={<AlertActionCloseButton onClose={() => closeSkipAlert()} />}
+        >
+          Skip status : {skipStatus}
```
Comment (cristianonicolai) [inline packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx]:
```suggestion
          Process execution failed to skip node which in error state. Message: {}
```

File: packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx [hunk 11]
```diff
@@ -77,8 +125,41 @@ const DataListItemComponent: React.FC<IOwnProps> = ({ id, instanceID, instanceSt
...
+        </Alert>)}
+        {retryAlertVisible && retryStatus === 'success' && (
+        <Alert
+            variant="success"
+            title="Success"
```
Comment (cristianonicolai) [inline packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx]:
```suggestion
            title="Retry operation"
```

File: packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx [hunk 12]
```diff
@@ -77,8 +125,41 @@ const DataListItemComponent: React.FC<IOwnProps> = ({ id, instanceID, instanceSt
...
+            action={<AlertActionCloseButton onClose={() => closeRetryAlert()} />}
+          >
+            Retry status : {retryStatus}
```
Comment (cristianonicolai) [inline packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx]:
```suggestion
            Process execution has successfully re executed node which was in error state.
```

File: packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx [hunk 13]
```diff
@@ -77,8 +125,41 @@ const DataListItemComponent: React.FC<IOwnProps> = ({ id, instanceID, instanceSt
...
+          </Alert>)}
+      {retryAlertVisible && retryStatus === 'failed' && (    
+          <Alert
+          variant="danger"
+          title="Failed"
+          action={<AlertActionCloseButton onClose={() => closeRetryAlert()} />}
+        >
+          Retry status : {retryStatus}
```
Comment (cristianonicolai) [inline packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx]:
```suggestion
          Process execution failed to re executed node which is error state. Message {}
```

**Review by cristianonicolai — CHANGES_REQUESTED** [review]

**Comment by cristianonicolai** [comment]
@Sara4994 @AjayJagan there is also an issue to open the dropdown of the last item in table. See screenshot below. Should I create a specific JIRA for that?

Selection_024

**Comment by cristianonicolai** [comment]
@Sara4994 also, when I click on retry or skip nothing happens.

File: packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx (hunk 10 above)
Comment (cristianonicolai) [reply packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx]: @mswiderski does that sounds accurate?

File: packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx (hunk 12 above)
Comment (cristianonicolai) [reply packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx]: @mswiderski does that sounds accurate?

File: packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx (hunk 13 above)
Comment (cristianonicolai) [reply packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx]: @mswiderski same here.

File: packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx (hunk 10 above)
Comment (mswiderski) [reply packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx]: sounds good to me

File: packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx (hunk 12 above)
Comment (mswiderski) [reply packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx]: sounds good to me

File: packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx (hunk 13 above)
Comment (mswiderski) [reply packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx]: sounds good, maybe just change `failed to re executed` to `failed to re execute`

Comment (Sara4994) [reply packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx]: added

File: packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx (hunk 12 above)
Comment (Sara4994) [reply packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx]: added

File: packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx (hunk 11 above)
Comment (Sara4994) [reply packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx]: added

File: packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx (hunk 10 above)
Comment (Sara4994) [reply packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx]: added

File: packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx [hunk 14]
```diff
@@ -77,8 +144,49 @@ const DataListItemComponent: React.FC<IOwnProps> = ({ id, instanceID, instanceSt
...
+          >
+            Process execution has successfully re executed node which was in error state.
+          </Alert>)}
+      {retryAlertVisible && retryError && (    
+          <Alert
+          variant="danger"
+          title="Retry operation"
+          action={<AlertActionCloseButton onClose={() => closeRetryAlert()} />}
+        >
+          Process execution failed to re executed node which is error state. Message {retryMessage}
```
Comment (cristianonicolai) [inline packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx]:
```suggestion
          Process execution failed to re execute node which is error state. Message {retryMessage}
```

File: packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx [hunk 15]
```diff
@@ -77,8 +144,49 @@ const DataListItemComponent: React.FC<IOwnProps> = ({ id, instanceID, instanceSt
...
+    await handleSkip(processID, instanceID, endpoint);
+  }
+
+  const handleRetryButton = async () => {
+    setOpenModal(!openModal);
+    await handleRetry(processID, instanceID, endpoint);
+  }
   return (
     <React.Fragment>
+      {skipAlertVisible && skipStatus === 'success' && (
```
Comment (cristianonicolai) [inline packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx]: @Sara4994 what if you define 3 variables: variant, title and message, that way you could have a single alert rendered?

File: packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx [hunk 16]
```diff
@@ -47,9 +61,53 @@ const DataListItemComponent: React.FC<IOwnProps> = ({ id, instanceID, instanceSt
...
+    setError(result.data);
+  },[])
+
+  const handleSkip = useCallback(async (_processID, _instanceID, _endpoint) => {
+    const processInstanceId = instanceID;
+    const processId = processID;
+    
+    try {
+      const result = await axios.get(`${endpoint}/management/process/${processId}/instances/${processInstanceId}/skip`);
+      setAlertType('success');
```
Comment (cristianonicolai) [inline packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx:83]:
```suggestion
      setAlertTitle('Skip operation');
      setAlertType('success');
```

File: packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx [hunk 17]
```diff
@@ -47,9 +61,53 @@ const DataListItemComponent: React.FC<IOwnProps> = ({ id, instanceID, instanceSt
...
+      setAlertMessage('Process execution has successfully skipped node which was in error state.');
+      setAlertVisible(true);
+    }
+    catch(error) {
+      setAlertType('danger');
```
Comment (cristianonicolai) [inline packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx:89]:
```suggestion
      setAlertTitle('Skip operation');
      setAlertType('danger');
```

File: packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx [hunk 18]
```diff
@@ -47,9 +61,53 @@ const DataListItemComponent: React.FC<IOwnProps> = ({ id, instanceID, instanceSt
...
+      setAlertVisible(true);
+    }
+  },[])
+  
+  const handleRetry = useCallback(async (_processID, _instanceID, _endpoint) => {
+    const processInstanceId = instanceID;
+    const processId = processID;
+    try {
+      const result = await axios.get(`${endpoint}/management/process/${processId}/instances/${processInstanceId}/retrigger`);
+      setAlertTitle('Skip operation');
```
Comment (cristianonicolai) [inline packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx]:
```suggestion
      setAlertTitle('Retry operation');
```

File: packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx [hunk 19]
```diff
@@ -47,9 +61,53 @@ const DataListItemComponent: React.FC<IOwnProps> = ({ id, instanceID, instanceSt
...
+      setAlertType('success');
+      setAlertMessage('Process execution has successfully re executed node which was in error state.');
+      setAlertVisible(true);
+    }
+    catch(error) {
+      setAlertTitle('Skip operation');
```
Comment (cristianonicolai) [inline packages/management-console/src/components/Molecules/DataListItemComponent/DataListItemComponent.tsx]:
```suggestion
      setAlertTitle('Retry operation');
```

**Review by cristianonicolai — CHANGES_REQUESTED** [review]

**Review by cristianonicolai — APPROVED** [review]

**Review by AjayJagan — APPROVED** [review]

**Pull request merged**